---
title: "Gitsbe Consensus Model Prediction Performance Analysis"
output:
  html_document:
    df_print: paged
---

## Intro

This notebook includes the ensemble model analysis performed on the 7500 
models generated by the `Gitsbe` module when running the DrugLogics 
computational pipeline for finding synergistic drug combinations. All these 
models were trained towards a consesus steady state signaling pattern. The 
input for the simulations and the output data are in the `consensus-2500` 
directory. The analysis will be presented step by step in the sections below.

The R version used for this analysis is: **3.4.4 (Someone to Lean On)**.

## Prerequisites

Firstly, we load the required libraries (you need to install them if you don't
have them):

```{r message=FALSE}
# require(RColorBrewer)
# require(heatmap.plus)
# require(gplots)
# require(rje)
```

and the relavant helper functions:
```{r}
# Set the working directory to the gitsbe-model-analysis folder: 
# setwd("pathTo/gitsbe-model-analysis")
source("Rscripts/input_functions.R")
source("Rscripts/output_functions.R")
source("Rscripts/analysis_functions.R")
#source("Rscripts/plot_functions.R")
```

## Input

**Two inputs** are used in this analysis:

- The consensus **steady state** file from ~1000 cell lines, which lists the 
nodes of the signaling network along with their (predicted) activity state. The
consensus steady state has been calculated using the gene expression and CNV 
data from all cell lines of the [CCLE](https://www.sevenbridges.com/ccle/) and
the [PARADIGM](https://doi.org/10.1093/bioinformatics/btq182) tool, while the 
majority rule was used to derive the activity state of each node - meaning that 
more than half 1's for one node in all cell lines would result in an activity 
state of 1 for that particualr node in the consensus steady state profile.
- The **models** directory, which is the same as the models directory produced 
by `Gitsbe` and has one `.gitsbe` file per model that includes this info:
    - The *fitness* to the consensus steady state (value between 0 and 1)
    - The *stable state* of the boolean model (can be either 1 stable state or 
none - the later rarely happens if not at all, since the models have been 
selected through a genetic evolution algorithm in `Gitsbe` and so in the end, 
only those with at least one stable state have higher fitness values and remain 
in the later generations)
```{r}
consensus.dir = "consensus-2500"
data.dir = paste(getwd(), "/", consensus.dir, "/", sep = "")
# results.dir = paste(data.dir, "results", sep = "")
# dir.create(results.dir, showWarnings = FALSE)
steady.state.file = paste(data.dir, "consensus_steady_state", sep = "")
models.dir = paste(data.dir, "models", sep = "")
```
Now, we parse the data into proper R objects. First, we get the full stable 
state per model:
```{r}
models = get.model.names(models.dir)
nodes  = get.node.names(models.dir)
models.stable.state = get.stable.state.from.models.dir(models.dir)

# example stable state of the first model (first 24 nodes)
head(models.stable.state[1, ], n = 24L)
```
```{r}
print(paste("Number of nodes: ", length(nodes), "Number of models: ", length(models)))
```
Next, we get the consensus steady state. Note that the data and tools used (e.g. 
PARADIGM, CCLE) to derive the activity states of the network nodes produce more 
nodes that are present in the consensus steady state vector but not in our 
network and so they need to be discarded. Also, we reorder the columns of the 
consensus vector so as to match the order of nodes as they are in the 
`models.stable.state` object.
```{r}
consensus.steady.state = get.consensus.steady.state(steady.state.file)
consensus.steady.state = 
  prune.to.common.nodes.and.reorder(consensus.steady.state, nodes)

# print the activity values of the first 24 nodes
head(consensus.steady.state, n = 24L)
```
We, also keep track of each model fitness to the steady state (a value closer 
to 1 corresponds to a better fit to the consensus steady state profile):
```{r}
models.fitness = get.fitness.from.models.dir(models.dir)
head(as.data.frame(models.fitness))
```




